### PR TITLE
Histogram compatibility with numpy 7364

### PR DESCRIPTION
**Pull request by chelseatroy**
The hist function should now work with either 'normed' or 'density' defined. 

If neither are defined, default is 'False.'
If both are defined as different things, the function raises an error.

Some tests added, and comment documentation updated, to reflect the changes.

### CODE REVIEW

File: determinism_tex.svg [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8" standalone="no"?>
```
Comment (QuLogic) [inline determinism_tex.svg]: Please rebase and remove this file entirely.

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -5875,10 +5875,10 @@ def table(self, **kwargs):
     #### Data analysis
 
     @_preprocess_data(replace_names=["x", 'weights'], label_namer="x")
-    def hist(self, x, bins=None, range=None, normed=False, weights=None,
+    def hist(self, x, bins=None, range=None, density=None, weights=None,
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: Changing the order is _probably_ not a good idea, for backwards compatibility's sake, even though people really shouldn't be expecting order-only arguments to work this far in.

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -5923,7 +5923,10 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
 
             Default is ``None``
 
-        normed : boolean, optional
+        normed or density : boolean, optional
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: `normed, density`

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -5932,23 +5935,28 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
...
+            will be treated as 'False.'
+
+            If both are set to different things, the hist function raises an
+            error.
 
         weights : (n, ) array_like or None, optional
             An array of weights, of the same shape as `x`.  Each value in `x`
             only contributes its associated weight towards the bin count
-            (instead of 1).  If `normed` is True, the weights are normalized,
+            (instead of 1).  If `normed` and/or 'density' is True, the weights are normalized,
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: Probably too long; wrap at 79 characters.

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -5932,23 +5935,28 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
...
             so that the integral of the density over the range remains 1.
 
             Default is ``None``
 
         cumulative : boolean, optional
             If `True`, then a histogram is computed where each bin gives the
             counts in that bin plus all bins for smaller values. The last bin
-            gives the total number of datapoints.  If `normed` is also `True`
+            gives the total number of datapoints.  If `normed` and/or 'density' is also `True`
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: Probably too long; wrap at 79 characters.

File: lib/matplotlib/axes/_axes.py [hunk 6]
```diff
@@ -5932,23 +5935,28 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
...
             then the histogram is normalized such that the last bin equals 1.
             If `cumulative` evaluates to less than 0 (e.g., -1), the direction
-            of accumulation is reversed.  In this case, if `normed` is also
+            of accumulation is reversed.  In this case, if `normed` and/or 'density' is also
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: Probably too long; wrap at 79 characters.

File: lib/matplotlib/axes/_axes.py [hunk 7]
```diff
@@ -6068,6 +6076,15 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
         .. plot:: mpl_examples/statistics/histogram_demo_features.py
 
         """
+
+        # This sets the density variable, if necessary, to its predecessor, 'normed.'
+        if density != None and normed != None and density != normed:
+            raise ValueError('The density and normed arguments represent the same concept. Please set only one of them, or set them to the same value.')
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: Definitely too long; wrap at 79 characters.

File: lib/matplotlib/axes/_axes.py [hunk 8]
```diff
@@ -6068,6 +6076,15 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
         .. plot:: mpl_examples/statistics/histogram_demo_features.py
 
         """
+
+        # This sets the density variable, if necessary, to its predecessor, 'normed.'
+        if density != None and normed != None and density != normed:
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: Use `is not None`.

File: lib/matplotlib/axes/_axes.py [hunk 9]
```diff
@@ -6068,6 +6076,15 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
...
+            raise ValueError('The density and normed arguments represent the same concept. Please set only one of them, or set them to the same value.')
+        elif normed != None and density == None:
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: Use `is None`.

File: lib/matplotlib/tests/test_axes.py [hunk 10]
```diff
@@ -2541,6 +2541,18 @@ def test_hist_stacked_normed():
     ax = fig.add_subplot(111)
     ax.hist((d1, d2), stacked=True, normed=True)
 
+@image_comparison(baseline_images=['hist_stacked_normed'])
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Two blank lines above.

File: lib/matplotlib/tests/test_axes.py [hunk 11]
```diff
@@ -3262,10 +3274,6 @@ def test_specgram_angle_freqs():
                       noverlap=noverlap, pad_to=pad_to, sides='onesided',
                       mode='phase', scale='dB')
 
-    with pytest.raises(ValueError):
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Why is this removed?

File: lib/matplotlib/tests/test_axes.py [hunk 12]
```diff
@@ -2541,6 +2541,18 @@ def test_hist_stacked_normed():
     ax = fig.add_subplot(111)
     ax.hist((d1, d2), stacked=True, normed=True)
 
+@image_comparison(baseline_images=['hist_stacked_normed'])
+def test_hist_stacked_density():
+    # make some data
+    d1 = np.linspace(1, 3, 20)
+    d2 = np.linspace(0, 10, 50)
+    fig = plt.figure()
+    ax = fig.add_subplot(111)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: `fig, ax = plt.subplots()`

**Comment by QuLogic** [comment]
Mostly minor formatting things; it seems to make sense otherwise.

**Comment by NelleV** [comment]
If that makes your life easier, we are using the tool `pep8` to check for coherent styles in matplotlib. You can install it and run it locally on your computer, and maybe even configure your text editor to display the pep8 violations.

File: lib/matplotlib/axes/_axes.py (hunk 2 above)
Comment (chelseatroy) [inline lib/matplotlib/axes/_axes.py]: Hi @QuLogic!

My understanding is that if a user is expecting order-only arguments to work this far in, then their argument will just take the name density instead of normed now. And since the name density is meant, in this change, to actually replace normed, the code will still be backward compatible. If the developer specifically named it normed, then the fact that it's named allows it to work even though normed is at the end now. 

Please let me know if this is inaccurate.

File: determinism_tex.svg (hunk 1 above)
Comment (chelseatroy) [inline determinism_tex.svg]: Hi @QuLogic!

Thanks for the quick response. I rebased, but my changes are up to date with the current remote. The contents of determinism_tex.svg appear to be tested in test_backend_svg.py (line 204). Would you still like me to remove it?

File: lib/matplotlib/axes/_axes.py (hunk 2 above)
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: I think this is safe to do.  If people are using positional it just works, if they are using keyword it still just works.  By putting this here we make it clearer which the preferred name is.

File: lib/matplotlib/axes/_axes.py [hunk 13]
```diff
@@ -5932,23 +5935,28 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
             of observations. If `stacked` is also `True`, the sum of the
             histograms is normalized to 1.
 
-            Default is ``False``
+            Default is ``None`` for both 'normed' and 'density.' If either is
+            set, then that value will be used. If neither are set, then the args
+            will be treated as 'False.'
+
+            If both are set to different things, the hist function raises an
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: I suggest 'if both are set, raise' even if they are the same.

File: lib/matplotlib/axes/_axes.py (hunk 3 above)
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: suggest putting 'density' first as that will (eventually) be the preferred name.

File: determinism_tex.svg (hunk 1 above)
Comment (QuLogic) [inline determinism_tex.svg]: This was added in your first commit. You'll need to rebase and edit the first commit to remove the file.

File: lib/matplotlib/tests/test_axes.py (hunk 11 above)
Comment (chelseatroy) [inline lib/matplotlib/tests/test_axes.py]: @QuLogic it shouldn't have been. My mistake—good catch!

**Comment by chelseatroy** [comment]
Pushed several changes as per @QuLogic's observations and suggestions!

**Comment by chelseatroy** [comment]
Force pushed to remove determinism_tex.svg from 3 commits back.

File: lib/matplotlib/axes/_axes.py (hunk 13 above)
Comment (chelseatroy) [inline lib/matplotlib/axes/_axes.py]: Change made and pushed.

**Comment by dstansby** [comment]
There are still a few PEP 8 violations that are causing the tests to fail (https://travis-ci.org/matplotlib/matplotlib/jobs/193148586#L1777), would it be okay to fix them?

They look like they are all lines that are longer than 79 characters, which is the maximum allowed line length.

File: lib/matplotlib/axes/_axes.py [hunk 14]
```diff
@@ -5932,24 +5935,31 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
...
+            error.
 
         weights : (n, ) array_like or None, optional
             An array of weights, of the same shape as `x`.  Each value in `x`
             only contributes its associated weight towards the bin count
-            (instead of 1).  If `normed` is True, the weights are normalized,
-            so that the integral of the density over the range remains 1.
+            (instead of 1).  If `normed` and/or 'density' is True,
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: should be just 'or'

File: lib/matplotlib/axes/_axes.py [hunk 15]
```diff
@@ -5932,24 +5935,31 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
...
 
         cumulative : boolean, optional
             If `True`, then a histogram is computed where each bin gives the
             counts in that bin plus all bins for smaller values. The last bin
-            gives the total number of datapoints.  If `normed` is also `True`
-            then the histogram is normalized such that the last bin equals 1.
-            If `cumulative` evaluates to less than 0 (e.g., -1), the direction
-            of accumulation is reversed.  In this case, if `normed` is also
-            `True`, then the histogram is normalized such that the first bin
+            gives the total number of datapoints. If `normed` and/or 'density'
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: only 'or'

File: lib/matplotlib/axes/_axes.py [hunk 16]
```diff
@@ -6032,7 +6042,7 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
         Returns
         -------
         n : array or list of arrays
-            The values of the histogram bins. See **normed** and **weights**
+            The values of the histogram bins. See **normed or density** and **weights**
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: This line is likely too long and needs to be wrapped.

File: lib/matplotlib/tests/test_axes.py [hunk 17]
```diff
@@ -2542,6 +2542,18 @@ def test_hist_stacked_normed():
     ax.hist((d1, d2), stacked=True, normed=True)
 
 
+@image_comparison(baseline_images=['hist_stacked_normed'])
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_axes.py]: If you are going to add an image test you also need to commit the baseline images.

If this adds coverage we don't have other wise the images (or at least the png) should be added, if not, just drop the decorator.

**Comment by tacaswell** [comment]
@chelseatroy ping on this.

**Comment by chelseatroy** [comment]
Hi @tacaswell. Just saw this. I guess for some reason I didn't receive a notification about your ping on this PR. If this still hasn't been done, I can make the changes this week...though I suspect someone has beat me to it.

**Comment by chelseatroy** [comment]
@tacaswell I guess I'm pretty lost as to what to do from here. All the build checks are failing, I don't really understand why, I basically just changed the character width of some comments.

**Comment by efiring** [comment]
Presumably because of other changes since the last time you added a commit, there are conflicts so that your changes no longer merge cleanly.  The "Resolve conflicts" button shows them nicely, and it looks like they will be easy to resolve.

**Comment by tacaswell** [comment]
```
=================================== FAILURES ===================================
 PEP8-check(ignoring E111 E114 E115 E116 E121 E122 E123 E124 E125 E126 E127 E128 E129 E131 E226 E240 E241 E242 E243 E244 E245 E246 E247 E248 E249 E265 E266 E704 W503) 
[gw0] linux -- Python 3.6.1 /home/travis/build/matplotlib/matplotlib/venv/bin/python
/home/travis/build/matplotlib/matplotlib/lib/matplotlib/axes/_axes.py:5874:80: E501 line too long (80 > 79 characters)
            set, then that value will be used. If neither are set, then the args
                                                                               ^
/home/travis/build/matplotlib/matplotlib/lib/matplotlib/axes/_axes.py:6013:1: W293 blank line contains whitespace
^
/home/travis/build/matplotlib/matplotlib/lib/matplotlib/axes/_axes.py:6014:80: E501 line too long (80 > 79 characters)
        # Sets the density variable, if necessary, to its predecessor, 'normed.'
                                                                               ^
!!!!!!!!!!!! xdist.dsession.Interrupted: stopping after 1 failures !!!!!!!!!!!!!
```

Looks like there are 2 formatting issues.

File: lib/matplotlib/axes/_axes.py (hunk 13 above)
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: The docs are out of sync with the code.

File: lib/matplotlib/axes/_axes.py [hunk 18]
```diff
@@ -5999,6 +6010,51 @@ def hist(self, x, bins=None, range=None, normed=False, weights=None,
...
+        # Sets the density variable, if necessary, to its predecessor, 'normed.'
+        if density is not None and normed is not None:
+            raise ValueError('The density and normed arguments represent the '
+                             'same concept. Please set only one of them.')
+        elif normed is not None and density is None:
+            density = normed
+        elif normed is None and density is None:
+            density = False
+
+        def _normalize_input(inp, ename='input'):
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: I am confused by this?  Did this come from the merge?

**Comment by dstansby** [comment]
Hi @chelseatroy, thanks for the PR! Since there were two PRs I have merged them together with a few fixes and put them in #8993 - you will still be attributed to the work when it gets merged though.
